nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py

```diff
@@ -360,50 +360,70 @@ def fwd_output_and_loss_func(dataloader_iter, model):
-                        dec_input_to_1024 = self.convert_tokens_to_range(dec_input[0, :, 0:audio_len])
-                        dec_input_wav = self.additional_models['encodec'].decode([[dec_input_to_1024[None], None]])[
-                            0, 0
-                        ]
-                        self.logger.experiment.add_audio("Target Wav", label_wav, self.global_step, 24000)
-                        self.logger.experiment.add_audio("Dec Input Wav", dec_input_wav, self.global_step, 24000)
-
-                        input_token_list = [
-                            context_and_question_tokens[0, 0, i].item()
-                            for i in range(context_and_question_tokens.shape[2])
-                        ]
-                        input_token_list = [t for t in input_token_list if t != 0 and t < 30000]
-                        input_text = self.frozen_model.tokenizer.ids_to_text(input_token_list)
-                        self.logger.experiment.add_text("Input Text", input_text, self.global_step)
-
-                        token_logits = out_logits[0]
-                        speech_logits_list = out_logits[1]
-                        if self.frozen_model.enc_dec_model.parallel_output:
-                            # Gather from tensor parallel region
-                            token_logits = tensor_parallel.gather_from_tensor_model_parallel_region(token_logits)
-                            for _i in range(len(speech_logits_list)):
-                                speech_logits_list[_i] = tensor_parallel.gather_from_tensor_model_parallel_region(
-                                    speech_logits_list[_i]
-                                )
-                        speech_logits = torch.stack(speech_logits_list, dim=-1)  # (t, b, 1024, 7)
-                        token_logits_example = token_logits[:, 0, :] * 1
-                        speech_logits_example = speech_logits[:, 0, :, :] * 1
-                        first_layer_tokens = token_logits_example.argmax(dim=1) - 30000
-                        other_layer_tokens = []
-                        for _i in range(speech_logits_example.shape[2]):
-                            other_layer_tokens.append(speech_logits_example[:, :, _i].argmax(dim=1))
-
-                        all_layer_tokens = torch.stack([first_layer_tokens] + other_layer_tokens)  # (8, t)
-                        all_layer_tokens = self.convert_tokens_to_range(
-                            all_layer_tokens, apply_offset_correction=False
-                        )
-                        all_layer_tokens = torch.clip(all_layer_tokens, 0, 1023)
-                        predicted_wav = self.additional_models['encodec'].decode([[all_layer_tokens[None], None]])[
-                            0, 0
-                        ]
-                        self.logger.experiment.add_audio("Pred Wav", predicted_wav, self.global_step, 24000)
+                        if torch.count_nonzero(speech_mask) == 0:
```

Comment:
When we train on both ASR and TTS, some rows in speech_mask will be zeros and some will be ones. We should consider slicing the batch into text outputs and speech outputs (as per speech_mask) and then log wer or audio. Another option is to compute WER only for batch indices where speech mask is zero and log audio for any index where speech_mask is one.

Reply:
This PR is just to add ASR as another SFT task and not a joint task. Currently, we need train_task to be either 'tts' or 'asr'. We cannot do joint training (asr and tts in the same batch) right now, what we can do is train for tts in one run and use the saved checkpoint to do ASR and so on. Having joint training (asr and tts in the same batch) might be trickier and might need changes to the dataloader as well.